### PR TITLE
mohu - Technical Training

### PR DESCRIPTION
**Pull request by mizosoft**

### CODE REVIEW

File: estate/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+    'name': 'RealEstate',
+    'version': '0.1',
+    'summary': 'Tracks RealEstate property and buyer offers',
+    'description': "Tracks RealEstate property and buyer offers",
+    'depends': [
+        'base'
+    ],
+    'installable': True,
+    'application': True,
+    'auto_install': False
```
Comment (vchu-odoo) [inline estate/__manifest__.py:11]: `installable` and `auto_install` are by default these values, no need to repeat it.
```suggestion
    'depends': ['base'],
    'application': True,
```

**Review by vchu-odoo — COMMENTED** [review]
Good work!

Only one small nitpicking and suggestion.

File: estate/models/estate_property.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+    property_tag_ids = fields.Many2many('estate.property.tag', string='Property Tags')
+
+    offer_ids = fields.One2many('estate.property.offer', 'property_id', string='Offers')
+
+    total_area = fields.Integer('Total Area (sqm)', compute='_compute_total_area')
+
+    best_offer = fields.Float('Best Offer', compute='_compute_best_offer')
+
+    accepted_offer_id = fields.Many2one('estate.property.offer', string='Accepted Offer')
```
Comment (vchu-odoo) [inline estate/models/estate_property.py:53]: This field is not mentioned in the tutorial.

File: estate/views/estate_property_views.xml [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        <field name="name">Property Types</field>
+        <field name="res_model">estate.property.type</field>
+        <field name="view_mode">tree,form</field>
+    </record>
+
+    <record id="estate_property_tag_action" model="ir.actions.act_window">
+        <field name="name">Property Types</field>
+        <field name="res_model">estate.property.tag</field>
+        <field name="view_mode">tree,form</field>
+    </record>
```
Comment (vchu-odoo) [inline estate/views/estate_property_views.xml:22]: for each model we create separate files like `estate_property_tag_view` and `estate_property_type_view` to organise code better and put there all views related to the model

File: estate/views/estate_property_views.xml [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+    <record id="estate_property_view_form" model="ir.ui.view">
+        <field name="name">estate.property.form</field>
+        <field name="model">estate.property</field>
+        <field name="arch" type="xml">
+            <form string="Property">
+                <header>
+                    <button name="action_mark_sold" type="object" string="Sold"/>
+                    <button name="action_mark_cancelled" type="object" string="Cancel"/>
+                </header>
+
```
Comment (vchu-odoo) [inline estate/views/estate_property_views.xml:90]: No extra lines needed inside templates/records.
```suggestion
```

File: estate/views/estate_property_views.xml [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+                    <group>
+                        <field name="status"/>
+                        <field name="property_type_id"/>
+                        <field name="postcode"/>
+                        <field name="date_availability"/>
+                        <field name="expected_price"/>  
+                        <field name="selling_price"/>
+                        <field name="best_offer"/>
+                    </group>
```
Comment (vchu-odoo) [inline estate/views/estate_property_views.xml:113]: does not correspond to the description in tutorial

File: estate/views/estate_property_views.xml [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+            <search string="Search Properties">
+                <field name="name"/>
+                <field name="postcode"/>
+                <field name="expected_price"/>
+                <field name="bedrooms"/>
+                <field name="living_area"/>
+                <field name="facades"/>
+                <field name="property_type_id"/>
+
+                <filter string="Available" name="available" domain="['|', ('status', '=', 'new'), ('status', '=', 'offer_received')]"/>
```
Comment (vchu-odoo) [inline estate/views/estate_property_views.xml:211]: Just an alternative way
```suggestion
                <filter string="Available" name="available" domain="[('status', 'in', ('new', 'offer_received'))]"/>
```

**Review by vchu-odoo — COMMENTED** [review]
Good job!
Continue like that :+1: 

There are some fields in estate property that do not correspond to the task description in the tutorial, is it intended?

File: estate/models/estate_property.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+    @api.depends('living_area', 'garden_area')
+    def _compute_total_area(self):
+        for record in self:
+            record.total_area = record.living_area + record.garden_area
+
+    @api.depends('offer_ids.price')
+    def _compute_best_offer(self):
+        for record in self: 
+            record.best_offer = max(self.offer_ids.mapped('price')) if len(self.offer_ids) > 0 else None
```
[the remark this answers is not in the document]
Comment (mizosoft) [reply estate/models/estate_property.py:69]: This is nice!

File: estate/views/estate_property_views.xml (hunk 6 above)
Comment (mizosoft) [reply estate/views/estate_property_views.xml:211]: Good suggestion!

File: estate/views/estate_property_views.xml (hunk 3 above)
Comment (mizosoft) [reply estate/views/estate_property_views.xml:22]: makes sense

File: estate/models/estate_property.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+class EsateProperty(models.Model):
+    _name = 'estate.property'
+    _description = 'Defines a real estate property'
+
+    name = fields.Char('Title', required=True, index=True)
+    description = fields.Text('Description')
+    postcode = fields.Char('Postcode')
+    date_availability = fields.Date('Available From', copy=False, 
+                                    default=lambda _: date.today() + relativedelta(months=3))
```
[the remark this answers is not in the document]
Comment (mizosoft) [reply estate/models/estate_property.py:16]: I knew about this but I noticed that `fields.Date` is just a wrapper over `datetime.date`, so I opted for the method that was more succinct. Do you think the second way is more appropriate.

**Comment by mizosoft** [comment]
> There are some fields in estate property that do not correspond to the task description in the tutorial, is it intended?

Yes, I added `accepted_offer_id` to allow a NO-OP to handle accepting or refusing an already accepted offer. I've modified the code to rely on `estate_property.selling_price` & `estate_property_offer.status` instead for these cases.

File: estate/views/estate_property_views.xml (hunk 4 above)
Comment (mizosoft) [reply estate/views/estate_property_views.xml:41]: answered below

File: estate/models/estate_property.py (hunk 2 above)
Comment (mizosoft) [reply estate/models/estate_property.py]: answered below

File: estate/models/estate_property.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (vchu-odoo) [reply estate/models/estate_property.py:17]: We should leave lambda ofc, it is always better to use odoo framework methods for consistency, performance or security
```suggestion
                                    default=lambda _: fields.Date().add(fields.Date().today(), months=3)
```

**Review by vchu-odoo — COMMENTED** [review]
Congrats on completing python tutorial, great job!

**Pull request closed without merging**